doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,164 @@
+that are useful to store in Git, `cabal.project.local` is for short-lived, local experiments
+(like profiling) that, in general, shouldn't be committed to Git.
+
+These settings are used implicitly in the next ``cabal build`` command:
+The first setting ``profiling: True`` tells GHC
+to insert performance measuring code into your application in general,
+while the profiling setting ``profiling-detail: late-toplevel`` further instructs
+GHC to use so-called
+`late-cost-center profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
+and insert measuring code only after important optimisations have been applied to your application code.
+This reduces the performance slow-down of profiling itself and gives you more realistic measurements.
+Further in-depth information on profiling with GHC and its compiler options can be found in the
+`GHC profiling guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html>`__
+
+The ``cabal list-bin`` command is a convenient way to find the path to the built executable, which
+is needed to run the application with the right RTS flags as in the last command:
+``/path/to/my-app +RTS -pj -RTS``.
+
+Finally, when our application run finishes,
+a profiling JSON report (due to option ``-pj``) is written to a ``<app-name>.prof`` file,
+i.e. ``my-app.prof``, in the current directory.
+You can load this JSON report ``my-app.prof`` into a visualizer
+and look for performance bottlenecks. One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
+See the
+`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
+on how to optimize your code based on the profiling results afterwards.
+
+Different report formats can be generated by using different RTS flags. Some useful ones are:
+
+- ``-p`` for a GHC's own
+  `standard report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#cost-centres-and-cost-centre-stacks>`__
+  ``<app-name>.prof``, which can be visualized with `profiteur <https://github.com/jaspervdj/profiteur>`__
+  or `ghcprofview <https://github.com/portnov/ghcprofview-hs>`__.
+- ``-pj`` for a
+  `JSON report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#json-profile-format>`__
+  ``<app-name>.prof``, which can be visualized with `Speedscope <https://speedscope.app>`__.
+- ``-l-p`` for a binary
+  `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
+  ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
+  be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__
+  and visualized with `Speedscope <https://speedscope.app>`__.
```

Comment:
GHC manual mentions only ThredScope. I wonder if it's a good idea to mention a workflow that doesn't seem to have an official blessing. But on the other hand, I much prefer Speedscope to ThreadScope myself.

Reply:
Maybe the GHC doc is just not uptodate. i would like to keep mentioning speedscope.

Reply:
Speedscope and ThreadScope have very different goals and use cases. Also, Threadscope can't consume JSON files. But, yes, let's make sure the GHC and cabal docs are in accord.

Reply:
> Speedscope and ThreadScope have very different goals and use cases

@Mikolaj I'm not making stuff up, I swear! ThreadScope is what GHC manual suggests: https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#tracing Should we file an issue?

Reply:
I think that section suggests ThreadScope for events related to threads. Eventlogs have been repurposed also for events related to profiling and, additionally, they can be output in JSON and read into Speedscope. That's the bit relevant for profiling. Yes, let's file an issue, because this section seems outdated and based on the assumption the original functionality of eventlogs is the only one that exists.